### ISSUE TITLE
when Grab anti-aliasing application produces some errors

### ISSUE THREAD

**Issue by z00318884** [baldurk/renderdoc#1322]
## Description

when Grab anti-aliasing application produces some errors.
example:
GL_INVALID_FRAMEBUFFER_OPERATION error is reported when gldraw element is executed, and then the function glCheckFramebufferStatus was  called to get the error message: GL_FRAMEBUFFER_INCOMPLETE_MULTISAMPLE

**Comment by baldurk** [comment]
Hi there I'm currently away until the end of next week so I will look at this when I'm back.

I will need a repro case to investigate this as renderdoc does not produce those errors, the driver does. So I need more information about what is wrong and why this seems like a renderdoc bug not an application bug.

**Comment by baldurk** [comment]
As above, this information does not help find a problem. I will need a reproducible test case or more relevant information to investigate the problem.

**Comment by z00318884** [comment]
Sorry, for some reason I can't give you the corresponding case, but I seem to have found the problem. You can help confirm if there is a problem listed below.
gl_dispatch_table.h  ：
PFNGLRENDERBUFFERSTORAGEMULTISAMPLEPROC glRenderbufferStorageMultisample;    // aliases glRenderbufferStorageMultisampleEXT

but when call glRenderbufferStorageMultisampleEXT API，Renderdoc does not enter “WrappedOpenGL::glRenderbufferStorageMultisample“  this function.


If I add a WrappedOpenGL::glRenderbufferStorageMultisampleEXT  function same as “WrappedOpenGL::glRenderbufferStorageMultisample“  , the problem doesn't exist anymore.

**Comment by baldurk** [comment]
I will still need more information to test this. Please give more details like which platform, GPU and driver you're running on, the API (GLES or GL), and the version of RenderDoc.

I'm guessing this is on GLES because `glRenderbufferStorageMultisample` was core in GL 3.0, so I ran [this ARM sample](https://github.com/ARM-software/opengl-es-sdk-for-android/tree/master/samples/advanced_samples/MultisampledFBO) which uses `glRenderbufferStorageMultisampleEXT` and it captures and replays correctly on RenderDoc v1.2 as well as the latest code. I also tried just in case on a simple desktop GL sample using the function but it worked correctly there as well.

The code you quoted treats `glRenderbufferStorageMultisampleEXT` identically to `glRenderbufferStorageMultisample` because the functions have identical signatures and semantics. The only difference is the name so there is no need for separate implementations. This is true for many many functions in GL.

Assuming this is GLES, I'm not clear if this is running on a GLES 3 context (where `glRenderbufferStorageMultisample` is available) or GLES 2 where only the EXT variant is available. The code-path will be slightly different because RenderDoc will call into the core function if it's available, even if the application requests via the EXT name. I did find a minor bug here where the core hook would be returned either way, but that doesn't change anything functionally.

I unfortunately can't test on a GLES 2 device because all of my devices are GLES 3. I tried manually hacking the code to pretend that `glRenderbufferStorageMultisample` doesn't exist, but assuming that was a valid test it seemed to correctly handle the code using `glRenderbufferStorageMultisampleEXT`.

If you can't share your program that shows the problem can you try running the ARM sample above linked and see if it works. If that breaks we can try and diagnose why it's breaking on your device, and if that works it shows the problem is not isolated to just calling `glRenderbufferStorageMultisampleEXT`, so you could try and modify the sample to show your problem by comparing it to your code.

**Comment by z00318884** [comment]
Here is the log when capturing and the print I add.

**Comment by z00318884** [comment]
GLenum nRet = GL.glGetError();
  if (nRet != 0) {
      GLenum nRet2 = GL.glCheckFramebufferStatus(eGL_FRAMEBUFFER);
      RDCLOG("Process WrappedOpenGL::glDrawElementsInstanced() nRet2 = 0x%x, error=0x%x\n", nRet2, nRet);
  }

**Comment by z00318884** [comment]
RDOC 013843: [15:56:53]        gl_driver.cpp(1185) - Log     - Activating new GL context: ARM / Mali-G72 / OpenGL ES 3.2 v1.r14p0-00cet0.d976e5b2d78148ce6020c4eb3ff0a2a3
RDOC 013843: [15:56:53]        gl_common.cpp( 494) - Log     - Vendor checks for 32 (ARM / Mali-G72 / OpenGL ES 3.2 v1.r14p0-00cet0.d976e5b2d78148ce6020c4eb3ff0a2a3)
RDOC 013843: [15:56:53]        gl_common.cpp( 783) - Warning - Reading from the depth buffer using glReadPixels is not supported (GL_NV_read_depth)
RDOC 013843: [15:56:53]        gl_common.cpp( 787) - Warning - Reading from the stencil buffer using glReadPixels is not supported (GL_NV_read_stencil)
RDOC 013843: [15:56:53]        gl_common.cpp( 791) - Warning - Reading from the packed depth-stencil buffers using glReadPixels is not supported (GL_NV_read_depth_stencil)
RDOC 013843: [15:56:53]  gl_shader_funcs.cpp(1235) - Error   - Assertion failed: 'record' Couldn't identify object passed to function. Mismatched or bad GLuint? (program=1)
RDOC 013843: [15:56:54]  gl_shader_funcs.cpp(1235) - Error   - Assertion failed: 'record' Couldn't identify object passed to function. Mismatched or bad GLuint? (program=3)
RDOC 013843: [15:56:54]     android_hook.cpp( 785) - Log     - Refreshing android hooks...
RDOC 013843: [15:56:54]     android_hook.cpp( 787) - Log     - Refreshed
RDOC 013843: [15:56:54]        egl_hooks.cpp( 182) - Debug   - eglCreateContext:
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 3098: 2
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 30fb: 0
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 30fc: 0
RDOC 013843: [15:56:54]        gl_driver.cpp( 959) - Log     - Core context D1D51A40 created with attribs, sharing with context 00000000
RDOC 013843: [15:56:54]     android_hook.cpp( 785) - Log     - Refreshing android hooks...
RDOC 013843: [15:56:54]     android_hook.cpp( 787) - Log     - Refreshed
RDOC 013843: [15:56:54]        egl_hooks.cpp( 182) - Debug   - eglCreateContext:
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 3098: 2
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 30fb: 0
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 30fc: 0
RDOC 013843: [15:56:54]        gl_driver.cpp( 959) - Log     - Core context BA6047C0 created with attribs, sharing with context D1D51A40
RDOC 013843: [15:56:54]     android_hook.cpp( 785) - Log     - Refreshing android hooks...
RDOC 013843: [15:56:54]     android_hook.cpp( 787) - Log     - Refreshed
RDOC 013843: [15:56:54]        egl_hooks.cpp( 182) - Debug   - eglCreateContext:
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 3098: 2
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 30fb: 0
RDOC 013843: [15:56:54]        egl_hooks.cpp( 189) - Debug   - 30fc: 0
RDOC 013843: [15:56:54]        gl_driver.cpp( 959) - Log     - Core context BA604800 created with attribs, sharing with context 00000000
RDOC 013843: [15:56:55]        gl_common.cpp( 461) - Log     - Checking enabled extensions, running as OpenGL ES 3.2
RDOC 013843: [15:56:55]      gl_emulated.cpp(2066) - Log     - Emulating ARB_clear_buffer_object
RDOC 013843: [15:56:55]      gl_emulated.cpp(2081) - Log     - Emulating ARB_internalformat_query2
RDOC 013843: [15:56:55]      gl_emulated.cpp(2135) - Log     - Emulating EXT_direct_state_access
RDOC 013843: [15:56:55]      gl_emulated.cpp(2222) - Log     - Emulating ARB_direct_state_access
RDOC 013843: [15:56:55]        gl_driver.cpp(1185) - Log     - Activating new GL context: ARM / Mali-G72 / OpenGL ES 3.2 v1.r14p0-00cet0.d976e5b2d78148ce6020c4eb3ff0a2a3
RDOC 013843: [15:56:55]         gl_manager.h( 125) - Debug   - Removed 0/44 resources belonging to context/sharegroup D1D51A40
RDOC 013843: [15:56:55]         gl_manager.h( 125) - Debug   - Removed 0/44 resources belonging to context/sharegroup BA6047C0
RDOC 013843: [15:56:55]         gl_manager.h( 125) - Debug   - Removed 0/44 resources belonging to context/sharegroup FFFFF82E
RDOC 013843: [15:56:55]         gl_manager.h( 125) - Debug   - Removed 2/42 resources belonging to context/sharegroup BA604800
RDOC 013843: [15:56:55]         gl_manager.h( 125) - Debug   - Removed 17/25 resources belonging to context/sharegroup FFFFF82D
RDOC 013843: [15:56:55]     android_hook.cpp( 785) - Log     - Refreshing android hooks...
RDOC 013843: [15:56:55]     android_hook.cpp( 787) - Log     - Refreshed
RDOC 013843: [15:56:55]        egl_hooks.cpp( 182) - Debug   - eglCreateContext:
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 3098: 3
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 30fb: 2
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 30fc: 0
RDOC 013843: [15:56:55]        gl_driver.cpp( 959) - Log     - Core context BA604800 created with attribs, sharing with context 00000000
RDOC 013843: [15:56:55]     android_hook.cpp( 785) - Log     - Refreshing android hooks...
RDOC 013843: [15:56:55]     android_hook.cpp( 787) - Log     - Refreshed
RDOC 013843: [15:56:55]        egl_hooks.cpp( 182) - Debug   - eglCreateContext:
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 3098: 3
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 30fb: 2
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 30fc: 0
RDOC 013843: [15:56:55]        gl_driver.cpp( 959) - Log     - Core context BA6047C0 created with attribs, sharing with context BA604800
RDOC 013843: [15:56:55]     android_hook.cpp( 785) - Log     - Refreshing android hooks...
RDOC 013843: [15:56:55]     android_hook.cpp( 787) - Log     - Refreshed
RDOC 013843: [15:56:55]        egl_hooks.cpp( 182) - Debug   - eglCreateContext:
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 3098: 3
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 30fb: 2
RDOC 013843: [15:56:55]        egl_hooks.cpp( 189) - Debug   - 30fc: 0
RDOC 013843: [15:56:55]        gl_driver.cpp( 959) - Log     - Core context D1D51A40 created with attribs, sharing with context 00000000
RDOC 013843: [15:56:55]        gl_common.cpp( 461) - Log     - Checking enabled extensions, running as OpenGL ES 3.2
RDOC 013843: [15:56:55]      gl_emulated.cpp(2066) - Log     - Emulating ARB_clear_buffer_object
RDOC 013843: [15:56:55]      gl_emulated.cpp(2081) - Log     - Emulating ARB_internalformat_query2
RDOC 013843: [15:56:55]      gl_emulated.cpp(2135) - Log     - Emulating EXT_direct_state_access
RDOC 013843: [15:56:55]      gl_emulated.cpp(2222) - Log     - Emulating ARB_direct_state_access
RDOC 013843: [15:56:55]        gl_driver.cpp(1185) - Log     - Activating new GL context: ARM / Mali-G72 / OpenGL ES 3.2 v1.r14p0-00cet0.d976e5b2d78148ce6020c4eb3ff0a2a3
RDOC 013843: [15:56:55]        gl_common.cpp( 461) - Log     - Checking enabled extensions, running as OpenGL ES 3.2
RDOC 013843: [15:56:55]      gl_emulated.cpp(2066) - Log     - Emulating ARB_clear_buffer_object
RDOC 013843: [15:56:55]      gl_emulated.cpp(2081) - Log     - Emulating ARB_internalformat_query2
RDOC 013843: [15:56:55]      gl_emulated.cpp(2135) - Log     - Emulating EXT_direct_state_access
RDOC 013843: [15:56:55]      gl_emulated.cpp(2222) - Log     - Emulating ARB_direct_state_access
RDOC 013843: [15:56:55]        gl_driver.cpp(1185) - Log     - Activating new GL context: ARM / Mali-G72 / OpenGL ES 3.2 v1.r14p0-00cet0.d976e5b2d78148ce6020c4eb3ff0a2a3
RDOC 013843: [15:57:05]     android_hook.cpp( 785) - Log     - Refreshing android hooks...
RDOC 013843: [15:57:05]     android_hook.cpp( 787) - Log     - Refreshed
RDOC 013843: [15:57:05]        egl_hooks.cpp( 182) - Debug   - eglCreateContext:
RDOC 013843: [15:57:05]        egl_hooks.cpp( 189) - Debug   - 3098: 2
RDOC 013843: [15:57:05]        gl_driver.cpp( 959) - Log     - Core context E5F72B80 created with attribs, sharing with context 00000000
RDOC 013843: [15:57:05]        gl_common.cpp( 461) - Log     - Checking enabled extensions, running as OpenGL ES 3.2
RDOC 013843: [15:57:05]      gl_emulated.cpp(2066) - Log     - Emulating ARB_clear_buffer_object
RDOC 013843: [15:57:05]      gl_emulated.cpp(2081) - Log     - Emulating ARB_internalformat_query2
RDOC 013843: [15:57:05]      gl_emulated.cpp(2135) - Log     - Emulating EXT_direct_state_access
RDOC 013843: [15:57:05]      gl_emulated.cpp(2222) - Log     - Emulating ARB_direct_state_access
RDOC 013843: [15:57:05]        gl_driver.cpp(1185) - Log     - Activating new GL context: ARM / Mali-G72 / OpenGL ES 3.2 v1.r14p0-00cet0.d976e5b2d78148ce6020c4eb3ff0a2a3
RDOC 013843: [15:57:05]         gl_manager.h( 125) - Debug   - Removed 2/190 resources belonging to context/sharegroup E5F72B80
RDOC 013843: [15:57:05]         gl_manager.h( 125) - Debug   - Removed 17/173 resources belonging to context/sharegroup FFFFF82A

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8cd5, error=0x500

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8cd5, error=0x500

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1679) - Log     - Process WrappedOpenGL::glDrawElementsInstanced() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1679) - Log     - Process WrappedOpenGL::glDrawElementsInstanced() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1679) - Log     - Process WrappedOpenGL::glDrawElementsInstanced() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1679) - Log     - Process WrappedOpenGL::glDrawElementsInstanced() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1258) - Log     - Process WrappedOpenGL::glDrawElements() nRet2 = 0x8d56, error=0x506

RDOC 013843: [15:57:44]    gl_draw_funcs.cpp(1679) - Log     - Process WrappedOpenGL::glDrawElementsInstanced() nRet2 = 0x8d56, error=0x506

**Comment by baldurk** [comment]
That doesn't tell me anything more than your original report, as all this shows is that the framebuffer is incomplete at the draw but not how it happened. In fact the first two errors show a `GL_INVALID_ENUM` error with a `GL_FRAMEBUFFER_COMPLETE` framebuffer status, which must be a different problem. It also looks like there are a couple of invalid calls to `glProgramParameteri` earlier.

Please see my previous comment for what I found and how this bug could be investigated further, if you can't share your program then by testing the linked ARM sample.

**Comment by z00318884** [comment]
I heard that ARM's glRenderbufferStorageMultisampleEXT is different from Qualcomm, and my application uses renderdoc on Qualcomm phones without problems. In the ARM mobile phone, you need to add glRenderbufferStorageMultisampleEXT to be no problem. But replay, I don't know exactly where to change, no modification, it will be the same mistake as the pre-report modification.

**Comment by baldurk** [comment]
I'm sorry but I don't understand your comment. `glRenderbufferStorageMultisampleEXT` is provided by `GL_EXT_multisampled_render_to_texture` on GLES, which is a standard extension across all vendors. In GLES 3.0 it was promoted to core, so it's also available as `glRenderbufferStorageMultisample` on all 3.0 drivers, but the function is identical.

It will have a different driver implementation, but as far as RenderDoc is concerned unless there is a platform-specific hooking problem then that doesn't matter. If you have errors about hooks failing in your logcat then there's nothing that can be done to solve the problem, this is just a limitation of Android.

**Comment by z00318884** [comment]
I mean the same application, the same renderdoc, only the driver platform is different. When the renderdoc is running in the background state on the mobile phone of the Mali chip platform, the application will have a partial black screen. The renderdoc runs on the Qualcomm chip platform mobile phone. In any state, the application is running normally.
So is this API for "glRenderbufferStorageMultisampleEXT", some modifications made in the mali driver, causing this problem?

**Comment by z00318884** [comment]
I just learned that in the Mali driver, glRenderbufferStorageMultisampleEXT performs the average processing on the buffer, and glRenderbufferStorageMultisample is the separate surface processing on the buffer.Does this have any effect?

**Comment by z00318884** [comment]
https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_multisampled_render_to_texture.txt

**Comment by baldurk** [comment]
I looked into this specifically on ARM vs Qualcomm and it looks like it is a driver bug. I've added a workaround in that commit which should hopefully fix the problem, by treating the functions as entirely separate instead of aliases.

Talking to ARM it seems like their driver doesn't properly handle FBOs that share attachments between the core functions and the equivalent extension functions, though it probably should. Any fix would take a long time to reach users though so the workaround should fix it.
